Index: settings
===================================================================
--- settings
+++ settings
@@ -1,20 +1,19 @@
 {
   "remappings": [
-    "@openzeppelin/contracts-upgradeable/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/contracts/",
-    "@openzeppelin/contracts/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/",
+    "@openzeppelin/contracts-upgradeable/=lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/contracts/",
+    "@openzeppelin/contracts/=lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/",
     "aave-v3-origin/=lib/aave-v3-origin/src/",
     "aave-v3-origin-tests/=lib/aave-v3-origin/tests/",
     "ds-test/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/forge-std/lib/ds-test/src/",
-    "erc4626-tests/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
+    "erc4626-tests/=lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
     "forge-std/=lib/forge-std/src/",
-    "openzeppelin-contracts-upgradeable/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/",
-    "openzeppelin-contracts/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/",
-    "solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "openzeppelin-contracts-upgradeable/=lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/",
+    "openzeppelin-contracts/=lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/",
+    "solidity-utils/=lib/aave-v3-origin/lib/solidity-utils/src/",
     "aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
     "aave-helpers/=lib/aave-helpers/",
-    "aave-v3-core/=lib/aave-v3-origin/src/core/",
-    "aave-v3-periphery/=lib/aave-v3-origin/src/periphery/"
+    "halmos-cheatcodes/=lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/halmos-cheatcodes/src/"
   ],
   "optimizer": {
     "enabled": true,
     "runs": 200
@@ -39,29 +38,29 @@
   "evmVersion": "shanghai",
   "viaIR": false,
   "libraries": {
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/BorrowLogic.sol": {
-      "BorrowLogic": "0x4c52FE2162200bf26c314d7bbd8611699139d553"
+      "BorrowLogic": "0x62325c94E1c49dcDb5937726aB5D8A4c37bCAd36"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/BridgeLogic.sol": {
-      "BridgeLogic": "0x97dCbFaE5372A63128F141E8C0BC2c871Ca5F604"
+      "BridgeLogic": "0x621Ef86D8A5C693a06295BC288B95C12D4CE4994"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
-      "ConfiguratorLogic": "0x3a593A622754ed9572599D33Aad6D799B0899Fae"
+      "ConfiguratorLogic": "0x09e88e877B39D883BAFd46b65E7B06CC56963041"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/EModeLogic.sol": {
-      "EModeLogic": "0x88F864670De467aA73CD45325F9652C578C8AB85"
+      "EModeLogic": "0xC31d2362fAeD85dF79d0bec99693D0EB0Abd3f74"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
-      "FlashLoanLogic": "0xb32381feFFF45eE9F47fD2f2cF83C832637d6EF0"
+      "FlashLoanLogic": "0x34039100cc9584Ae5D741d322e16d0d18CEE8770"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
-      "LiquidationLogic": "0x80d16970B31243Fe67DaB028115f3E4c3E3510Ad"
+      "LiquidationLogic": "0x4731bF01583F991278692E8727d0700a00A1fBBf"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/PoolLogic.sol": {
-      "PoolLogic": "0xA58FB47bE9074828215A173564C0CD10f6F249bf"
+      "PoolLogic": "0xf8C97539934ee66a67C26010e8e027D77E821B0C"
     },
     "lib/aave-v3-origin/src/contracts/protocol/libraries/logic/SupplyLogic.sol": {
-      "SupplyLogic": "0x2b22E425C1322fbA0DbF17bb1dA25d71811EE7ba"
+      "SupplyLogic": "0x185477906B46D9b8DE0DEB73A1bBfb87b5b51BC3"
     }
   }
 }
\ No newline at end of file
