===================================================================
--- src/contracts/instances/PoolConfiguratorInstance.sol
+++ lib/aave-v3-origin/src/contracts/instances/PoolConfiguratorInstance.sol
@@ -3,9 +3,9 @@
 
 import {PoolConfigurator, IPoolAddressesProvider, IPool, VersionedInitializable} from "../protocol/pool/PoolConfigurator.sol";
 
 contract PoolConfiguratorInstance is PoolConfigurator {
-    uint256 public constant CONFIGURATOR_REVISION = 4;
+    uint256 public constant CONFIGURATOR_REVISION = 5;
 
     /// @inheritdoc VersionedInitializable
     function getRevision() internal pure virtual override returns (uint256) {
         return CONFIGURATOR_REVISION;
