===================================================================
--- src/contracts/interfaces/IPoolDataProvider.sol
+++ lib/aave-v3-origin/src/contracts/interfaces/IPoolDataProvider.sol
@@ -255,5 +255,12 @@
      */
     function getVirtualUnderlyingBalance(
         address asset
     ) external view returns (uint256);
+
+    /**
+     * @notice Returns the deficit of the reserve
+     * @param asset The address of the underlying asset of the reserve
+     * @return The reserve deficit
+     */
+    function getReserveDeficit(address asset) external view returns (uint256);
 }
