Index: BridgeLogic.sol
===================================================================
--- BridgeLogic.sol
+++ BridgeLogic.sol
@@ -13,8 +13,13 @@
 import {Errors} from "../helpers/Errors.sol";
 import {ValidationLogic} from "./ValidationLogic.sol";
 import {ReserveLogic} from "./ReserveLogic.sol";
 
+/**
+ * @title BridgeLogic library
+ * @author Aave
+ * @notice Implements functions to mint unbacked aTokens and back the unbacked tokens.
+ */
 library BridgeLogic {
     using ReserveLogic for DataTypes.ReserveCache;
     using ReserveLogic for DataTypes.ReserveData;
     using UserConfiguration for DataTypes.UserConfigurationMap;
